### PR TITLE
Docs: render autodoc & typehints in the API documentation

### PR DESCRIPTION
**Pull request by anilbey**
## Description

Docs now show all of the public functions in the API. Before it was only listing their names.
Docs show the argument types and return types. API docs are type checked using type hints and a static type checker to ensure they are always up-to-date.
Docs show CHANGELOG.
A few fixes in the function/method docstrings.

### Before

Only the function names and the first line of docstring are provided.

docs-screenshot-before

### After

Not only the first line but the complete docstring get generated for each function. Return types and argument types are annotated using type hints (instead of using docstrings).

docs-screenshot-after

## Checklist:
- [X] Unit tests are added to cover the changes.
- [X] The changes are mentioned in the documentation.
- [X] CHANGELOG.md is updated (skip if the change is not important for the changelog).

### CODE REVIEW

File: efel/api.py [hunk 1]
```diff
@@ -343,41 +332,33 @@ def getFeatureValues(
...
+        parallel_map: Map function to parallelise over the traces. Default is the
+                      serial map() function
+        return_list: By default the function returns a list of dicts. This
+                     optional argument can disable this, so that the result of the
+                     parallel_map() is returned. Can be useful for performance
+                     reasons when an iterator is preferred.
+        raise_warnings: Raise warning when efel c++ returns an error
+
+    Returns:
+        For every input trace a feature value dict is return (in
```
Comment (AurelienJaquier) [inline efel/api.py:349]: typo: return -> returned

File: efel/api.py [hunk 2]
```diff
@@ -428,55 +419,59 @@ def _get_feature_values_serial(trace_featurenames):
...
-    feature_values : list of dicts
-                     For every input trace a feature value dict is return (in
-                     the same order). The dict contains the keys of
-                     'feature_names', every key contains the mean of the array
-                     that is returned by getFeatureValues()
-                     The value is None if an error occured during the
-                     calculation of the feature, or if the feature value array
-                     was empty.
+    Args:
+        traces: Every trace dict represent one trace. The dict should have the
```
Comment (AurelienJaquier) [inline efel/api.py:440]: typo: represent -> represents

File: efel/api.py [hunk 3]
```diff
@@ -428,55 +419,59 @@ def _get_feature_values_serial(trace_featurenames):
...
+                following keys: 'T', 'V', 'stim_start', 'stim_end'
+        feature_names: List with the names of the features to be calculated on all
+                       the traces.
+        raise_warnings: Raise warning when efel c++ returns an error
+
+    Returns:
+        For every input trace a feature value dict is return (in
```
Comment (AurelienJaquier) [inline efel/api.py:447]: typo: is return -> is returned

File: efel/io.py [hunk 4]
```diff
@@ -37,27 +37,31 @@ def load_ascii_input(
...
-        "stimulus_end", "stimulation_end", "current_injection_end"
-
-        Returned objects
-        ====================
-        stim_start : numerical value (ms) or None
-        stim_end : numerical value (ms) or None
+    Seeks for the stim_start and stim_end parameters inside the Neo data.
+
+    Args:
+        blocks (Neo object blocks): Description of what blocks represents.
```
[the remark this answers is not in the document]
Comment (anilbey) [reply efel/io.py:45]: This function is written 7 year ago using the Neo package at that time. There is an internal efel ticket to update the usage to the recent one. There is a high chance that this function will be written in a much simpler way. Shall we update the args once that ticket it through?

[the remark this answers is not in the document]
Comment (AurelienJaquier) [reply efel/io.py:45]: Alright, sure, no problem

File: efel/io.py [hunk 5]
```diff
@@ -150,31 +154,32 @@ def extract_stim_times_from_neo_data(blocks, stim_start, stim_end):
...
-        "current_injection"
-        First Event.name should be "stim_start", "stimulus_start",
-        "stimulation_start", "current_injection_start"
-        Second Event.name should be one of "stim_end", "stimulus_end",
-        "stimulation_end", "current_injection_end"
-
-        The returned object is presented like this :
-            returned object : [Segments_1, Segments_2, ..., Segments_n]
-            Segments_1 = [Traces_1, Traces_2, ..., Traces_n]
+    Seeks for the stim_start and stim_end parameters inside the Neo data.
```
[the remark this answers is not in the document]
Comment (anilbey) [reply efel/io.py:159]: Oh good that you noticed. This docstring and the one above were both raising sphinx errors, I must have copied the wrong one.

File: efel/io.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ilkilic) [reply efel/io.py:45]: Noted, I will make the update when I work on this ticket

File: efel/pyfeatures/pyfeatures.py [hunk 6]
```diff
@@ -419,26 +420,24 @@ def spikes_in_burst1_burstlast_diff():
...
     if exitCode < 0:
         if raise_warnings:
-            import warnings
             warnings.warn(
-                "Error while calculating feature %s: %s" %
-                (featureName, cppcore.getgError()),
+                f"Error while calculating {feature_name}, {cppcore.getgError()}",
                 RuntimeWarning)
         return None
     else:
```
[the remark this answers is not in the document]
Comment (anilbey) [reply efel/pyfeatures/pyfeatures.py:439]: Coming with the next push.

File: efel/api.py [hunk 7]
```diff
@@ -1,10 +1,10 @@
+from __future__ import annotations
```
[the remark this answers is not in the document]
Comment (anilbey) [reply efel/api.py:1]: It is being updated in the next push

File: efel/api.py [hunk 8]
```diff
@@ -104,157 +89,138 @@ def reset():
...
+def setDependencyFileLocation(location: str | Path) -> None:
+    """Sets the location of the Dependency file.
 
-    Parameters
-    ==========
-    location : string
-               path to the location of a Dependency file
-    """
+    eFEL uses 'Dependency' files to let the user define versions of features to use.
+    The installation directory of the eFEL contains a default 'DependencyV5.txt' file.
```
[the remark this answers is not in the document]
Comment (anilbey) [reply efel/api.py:97]: Next push

File: efel/api.py [hunk 9]
```diff
@@ -104,157 +89,138 @@ def reset():
...
 
-    Returns
-    =======
-    feature_names : list of strings
-                    A list that contains all the feature names available in
-                    the eFEL. These names can be used in the featureNames
-                    argument of e.g. getFeatureValues()
+    Returns:
+        A list that contains all the feature names available in
+        the eFEL. These names can be used in the featureNames
```
[the remark this answers is not in the document]
Comment (anilbey) [reply efel/api.py:162]: Coming with the next push

File: efel/api.py [hunk 10]
```diff
@@ -343,41 +332,33 @@ def getFeatureValues(
...
-    =======
-    feature_values : list of dicts
-                     For every input trace a feature value dict is return (in
-                     the same order). The dict contains the keys of
-                     'feature_names', every key contains a numpy array with
-                     the feature values returned by the C++ efel code.
-                     The value is None if an error occured during the
-                     calculation of the feature.
+    Args:
+        traces: Every trace dict represent one trace. The dict should have the
```
[the remark this answers is not in the document]
Comment (anilbey) [reply efel/api.py:336]: :+1:

File: efel/api.py (hunk 3 above)
Comment (anilbey) [reply efel/api.py:447]: :+1:

File: efel/api.py (hunk 2 above)
Comment (anilbey) [reply efel/api.py:440]: :+1:

File: efel/api.py (hunk 1 above)
Comment (anilbey) [reply efel/api.py:349]: :+1:

File: efel/api.py [hunk 11]
```diff
@@ -104,157 +89,138 @@ def reset():
...
-               path to the location of a Dependency file
+    Returns:
+        Path to the location of a Dependency file.
     """
-
     return _settings.dependencyfile_path
 
 
-def setThreshold(newThreshold):
+def set_threshold(new_threshold: float) -> None:
```
Comment (ilkilic) [inline efel/api.py:122]: why change the function name here?

Comment (anilbey) [reply efel/api.py:122]: To make it Python style and also to make it consistent with the rest of Python modules.
https://pep8.org/#naming-conventions

**Review by AurelienJaquier — APPROVED** [review]
Looks good to me but you might have some additional changes to make after pulling master hehehe

**Review by AurelienJaquier — APPROVED** [review]
alright, looks good!

**Comment by anilbey** [comment]
@AurelienJaquier ah yes, mypy asked to handle the possible `None` values returned from `get_cpp_feature` for `phaseslope_max`.

**Review by ilkilic — APPROVED** [review]
Looks great to me!

**Pull request merged**